toolchain/sem_ir/dump.cpp

```diff
@@ -6,305 +6,305 @@
 
 #include "toolchain/sem_ir/dump.h"
 
-#include "common/ostream.h"
+#include "common/raw_string_ostream.h"
 #include "toolchain/sem_ir/stringify_type.h"
 
 namespace Carbon::SemIR {
 
-static auto DumpNameIfValid(const File& file, NameId name_id) -> void {
+static auto DumpNameIfValid(const File& file, NameId name_id) -> std::string {
+  RawStringOstream out;
   if (name_id.has_value()) {
-    llvm::errs() << " `" << file.names().GetFormatted(name_id) << "`";
+    out << " `" << file.names().GetFormatted(name_id) << "`";
   }
+  return out.TakeStr();
 }
 
-static auto DumpNoNewline(const File& file, ConstantId const_id) -> void {
-  llvm::errs() << const_id;
-  if (const_id.is_symbolic()) {
-    llvm::errs() << ": "
-                 << file.constant_values().GetSymbolicConstant(const_id);
-  } else if (const_id.is_concrete()) {
-    llvm::errs() << ": "
-                 << file.insts().Get(
-                        file.constant_values().GetInstId(const_id));
-  }
-}
-
-static auto DumpNoNewline(const File& file, InstId inst_id) -> void {
-  llvm::errs() << inst_id;
+static auto DumpInstSummary(const File& file, InstId inst_id) -> std::string {
+  RawStringOstream out;
+  out << inst_id;
   if (inst_id.has_value()) {
-    llvm::errs() << ": " << file.insts().Get(inst_id);
-  }
-}
-
-static auto DumpNoNewline(const File& file, InterfaceId interface_id) -> void {
-  llvm::errs() << interface_id;
-  if (interface_id.has_value()) {
-    const auto& interface = file.interfaces().Get(interface_id);
-    llvm::errs() << ": " << interface;
-    DumpNameIfValid(file, interface.name_id);
+    out << ": " << file.insts().Get(inst_id);
   }
+  return out.TakeStr();
 }
 
-static auto DumpNoNewline(const File& file, SpecificId specific_id) -> void {
-  llvm::errs() << specific_id;
-  if (specific_id.has_value()) {
-    llvm::errs() << ": " << file.specifics().Get(specific_id);
-  }
-}
-
-LLVM_DUMP_METHOD auto Dump(const File& file, ClassId class_id) -> void {
-  llvm::errs() << class_id;
+LLVM_DUMP_METHOD auto Dump(const File& file, ClassId class_id) -> std::string {
+  RawStringOstream out;
+  out << class_id;
   if (class_id.has_value()) {
     const auto& class_obj = file.classes().Get(class_id);
-    llvm::errs() << ": " << class_obj;
-    DumpNameIfValid(file, class_obj.name_id);
+    out << ": " << class_obj << DumpNameIfValid(file, class_obj.name_id);
   }
-  llvm::errs() << '\n';
+  return out.TakeStr();
 }
 
-LLVM_DUMP_METHOD auto Dump(const File& file, ConstantId const_id) -> void {
-  DumpNoNewline(file, const_id);
-  llvm::errs() << '\n';
+LLVM_DUMP_METHOD auto Dump(const File& file, ConstantId const_id)
+    -> std::string {
+  RawStringOstream out;
+  out << const_id;
+  if (const_id.is_symbolic()) {
```

Comment:
HEAD now handles correctly if const id is none, instead of DCHECK-ing in here

```suggestion
  if (!const_id.has_value()) {
    return out.TakeStr();
  }
  if (const_id.is_symbolic()) {
```

Reply:
This suggests you may need to (perhaps painfully) merge a number of other dump changes that landed in #5190

Reply:
What you're seeing is intentional; GitHub shows the conflicts. Rebases and merges are a frequent topic of discussion, and the general conclusion has been to do neither during review because of the issues it causes for review.

Would you rather review this change as-is, so that you can see the progression, or should I merge head?

Reply:
Okay, was noting the merge will be painful. It's been helpful to review before merging, we can go ahead and merge now.